src/adapter/src/coord.rs

```diff
@@ -2131,178 +2140,203 @@ impl Coordinator {
-        // our `as_of`.
+        // An exception are storage collections that have an `upper` that's less than their `since`
+        // (most likely because they have not yet produced their snapshot). For these collections
+        // we only need to provide output starting from their `since`s, so these serve as upper
+        // bounds for our `as_of`.
         let mut max_as_of = Antichain::new();
-        for mv_id in &dependent_matviews {
-            let since = self.storage_implied_capability(*mv_id);
-            let upper = self.storage_write_frontier(*mv_id);
+        for id in &storage_constraints.dependants {
+            let since = self.storage_implied_capability(*id);
+            let upper = self.storage_write_frontier(*id);
             max_as_of.meet_assign(&since.join(upper));
         }
 
+        // For compute reconciliation to recognize that an existing dataflow can be reused, we want
+        // to advance the `as_of` far enough that it is beyond the `as_of`s of all dataflows that
+        // might still be installed on replicas, but ideally not much farther as that would prevent
+        // dataflow warmup.
+        //
+        // The compute controller maintains warmup capabilities for each dataflow, at the meet of
+        // the greatest available read frontiers (i.e. `upper.step_back()`) of all dataflow inputs.
+        // Dataflows installed on replicas are thus prevented from compacting beyond these warmup
+        // frontiers, so if we can reconstruct them, we know that reconciliation will succeed. Note
+        // that to reconstruct a warmup frontier, we must only take into account the frontiers of
+        // (transitive) storage dependencies: The frontiers of index dependencies can regress
+        // across restarts, so we'd risk ending up with a warmup frontier that is too early to
+        // ensure successful reconciliation.
+        //
+        // Note that choosing the warmup frontier only based on storage dependencies means that the
+        // dataflow might have to wait for indexes in between to catch up to this frontier, if
+        // their `as_of` is selected earlier.
+        let storage_dependencies = storage_constraints.dependencies_bundle();
+        let warmup_frontier = self.greatest_available_read(&storage_dependencies);
+
+        // Apply additional constraints based on whether this dataflow exports an index or a
+        // materialized view.
+        let write_frontier;
+        let candidate_as_of = match dataflow.sink_exports.keys().at_most_one() {
+            // Materialized view dataflow.
+            Ok(Some(sink_id)) => {
+                // If the target storage collection of a materialized view is already sealed, there
+                // is no need to install a dataflow in the first place.
+                write_frontier = self.storage_write_frontier(*sink_id).clone();
+                if write_frontier.is_empty() {
+                    Antichain::new()
```

Comment:
The `max_as_of` should also be empty in this case, right? Maybe you could assert that here.

Reply:
Yep! In fact `write_frontier` and `max_as_of` should always be equal for MVs since an MV dataflow is always depended on only by its target storage collection. Added a `soft_assert`.